taichi/backends/dx/dx_api.cpp

```diff
@@ -0,0 +1,320 @@
+
+      hWnd = CreateWindowEx(0,           // Optional window styles.
+                            CLASS_NAME,  // Window class
+                            std::wstring(window_name.begin(), window_name.end())
+                                .data(),          // Window text
+                            WS_OVERLAPPEDWINDOW,  // Window style
+                            // Size and position
+                            CW_USEDEFAULT, CW_USEDEFAULT,
+                            window_rect.right - window_rect.left,
+                            window_rect.bottom - window_rect.top,
+                            NULL,                // Parent window
+                            NULL,                // Menu
+                            GetModuleHandle(0),  // Instance handle
+                            NULL                 // Additional application data
+      );
+      TI_ERROR_IF(hWnd == NULL, "Window creation failed");
+      ShowWindow(hWnd, SW_SHOWDEFAULT);
+    }
+    #endif
+
+    CreateComputeDevice(&g_device, &g_context, hWnd, pp_swapchain, false);
+
+    if (g_swapchain) {
+      g_swapchain->Present(0, 0);
+      g_swapchain->Present(0, 0);
+    }
+
+    TI_TRACE("Creating D3D11 buffers");
+    const int N = 1048576;
+    CreateStructuredBuffer(g_device, 4, N, nullptr, &g_data_i32_buf);
+    CreateStructuredBuffer(g_device, 4, N, nullptr, &g_data_f32_buf);
+    CreateStructuredBuffer(g_device, 4, N, nullptr, &g_args_i32_buf);
+    CreateStructuredBuffer(g_device, 4, N, nullptr, &g_args_f32_buf);
+    CreateStructuredBuffer(g_device, 4, N, nullptr, &g_extr_i32_buf);
+    CreateStructuredBuffer(g_device, 4, N, nullptr, &g_extr_f32_buf);
+    char *zeroes = new char[N * 4];
+    memset(zeroes, 0x00, N * 4);
+    CreateRawBuffer(g_device, 4 * N, zeroes, &g_locks_buf);
+    delete[] zeroes;
+
+    TI_TRACE("Creating D3D11 UAVs");
+    CreateBufferUAV(g_device, g_data_i32_buf, &g_data_i32_uav);
+    CreateBufferUAV(g_device, g_data_f32_buf, &g_data_f32_uav);
+    CreateBufferUAV(g_device, g_args_i32_buf, &g_args_i32_uav);
+    CreateBufferUAV(g_device, g_args_f32_buf, &g_args_f32_uav);
+    CreateBufferUAV(g_device, g_extr_i32_buf, &g_extr_i32_uav);
+    CreateBufferUAV(g_device, g_extr_f32_buf, &g_extr_f32_uav);
+    CreateBufferUAV(g_device, g_locks_buf, &g_locks_uav);
+
+    // Copy to the UAVs
+    D3D11_BUFFER_DESC desc;
+    g_args_f32_buf->GetDesc(&desc);
+    D3D11_BUFFER_DESC tmp_desc = {};
+    tmp_desc.ByteWidth = desc.ByteWidth;
+    tmp_desc.BindFlags = 0;
+    tmp_desc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
+    tmp_desc.MiscFlags = 0;
+    tmp_desc.Usage = D3D11_USAGE_STAGING;
+    HRESULT hr = g_device->CreateBuffer(&tmp_desc, nullptr, &tmp_arg_buf);
+    assert(SUCCEEDED(hr));
+
+  } else {
+    TI_TRACE("D3D11 device has already been created.");
+  }
+  return true;
+}
+
+bool is_dx_api_available() {
+  return initialize_dx();
+}
+
+std::unique_ptr<Device> get_dx_device() {
```

Comment:
This should really be called `make_dx_device()`

Reply:
Removed this function altogether, this was only used once in materialize_runtime, and now materialize_runtime just calls the ctor of DxDevice.